src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs

```diff
@@ -63,6 +78,22 @@ public async Task<string> GetSecretValue(Uri secretUri, CancellationToken cancel
             return secretValue;
         }
 
+        public bool ShouldRefreshKeyVaultSecrets()
+        {
+            // return true if the RefreshAt time of any cached secret has already elapsed.
+            return _cachedKeyVaultSecrets.Any(cachedSecret => cachedSecret.Value.RefreshAt.HasValue && cachedSecret.Value.RefreshAt.Value < DateTimeOffset.UtcNow);
```

Comment:
>_cachedKeyVaultSecrets [](http://example.com/codeflow?start=19&length=22)

Iterating over concurrent dictionary creates a copy each time. There's an optimization to be had in this provider around caching the next known expiration time.

Reply:
Tracking the next refresh time using a private field instead of iterating over concurrent dictionary.

Reply:
I misspoke. Only [`Keys`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.keys?view=net-5.0#System_Collections_Concurrent_ConcurrentDictionary_2_Keys) and [`Values`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.values?view=net-5.0#System_Collections_Concurrent_ConcurrentDictionary_2_Values) creates a copy.

[Enumeration ](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getenumerator?view=net-5.0#System_Collections_Concurrent_ConcurrentDictionary_2_GetEnumerator)uses the dictionary in place.

Reply:
Reverted.